features/exacte_match.feature

```diff
@@ -0,0 +1,147 @@
+# language: nl
+
+Functionaliteit: leveren van objecten die exact overeenkomen met een opgegeven huisnummer, eventueel in combinatie met een huisletter en/of huisnummertoevoeging
+
+Deze functionaliteit biedt de mogelijkheid om een specifiek object te bevragen ook wanneer er meerdere objecten aan de gegeven parameters voldoen.
+Parameter exacteMatch=true werkt als een filter op de objecten die voldoen aan de opgegeven parameters huisnummer, huisletter en huisnummertoevoeging.
+De endpoints waar exacte match kan worden toegepast kennen geen fuzzy search.
+
+Wanneer een zoekvraag met de opgegeven parameters meerdere resultaten oplevert en exacteMatch=false, dan worden alle zoekresultaten die overeenkomen met de zoekcriteria geretourneerd.
+
+Wanneer een zoekvraag met de opgegeven parameters meerdere resultaten oplevert en exacteMatch=true, dan wordt alleen het object dat exact overeenkomt met de zoekcriteria geretourneerd.
+Er zijn situaties waarbij er ondanks het gebruik van exacteMatch=true meerdere resultaten mogelijk zijn, in dat geval worden alle resultaten geretourneerd.
+
+De default waarde van de exacteMatch parameter is false.
+
+Scenario: standaard gedrag
+  Gegeven een verzameling van 3 objecten met postcode=1234AA 
+  En één object (1) met huisnummer=3
+  En geen huisletter
+  En geen huisnummertoevoeging
+  En één object (2) met huisnummer=3
+  En huisletter=A
+  En geen huisnummertoevoeging
+  En één object (3) met huisnummer=3
+  En huisletter=B
+  En geen huisnummertoevoeging
+  Als er wordt gezocht met postcode=1234AA 
+  En huisnummer=3
+  En geen huisletter
+  En geen huisnummertoevoeging 
+  En exacteMatch=false
+  Dan bevat het resultaat alle 3 de objecten
+	
+  Gegeven een verzameling van 3 objecten met postcode=1234AA 
+  En één object (1) met huisnummer=3
+  En geen huisletter
+  En geen huisnummertoevoeging
+  En één object (2) met huisnummer=3
+  En huisletter=A
+  En geen huisnummertoevoeging
+  En één object (3) met huisnummer=3
+  En huisletter=B
+  En geen huisnummertoevoeging
+  Als er wordt gezocht met postcode=1234AA 
+  En huisnummer=3
+  En geen huisletter
+  En geen huisnummertoevoeging 
+  En exacteMatch=true
+  Dan bevat het resultaat alleen het object (1) met postcode=1234AA 
+  En huisnummer=3
+
```

Comment:
Ik weet niet of je bekend bent met de tables functionaliteit van Gherkin, want daarmee kan je deze scenario's veel compacter en met minder herhalingen beschrijven. v.b.

Gegeven de volgende verzameling van objecten
| postcode | huisnummer | huisletter | huisnummertoevoeging |
| 1234AA   | 3          |            |                      |
| 1234AA   | 3          | A          |                      |
| 1234AA   | 3          | B          |                      |
Als er wordt gezocht met
| postcode | huisnummer | huisletter | huisnummertoevoeging | exacteMatch |
| 1234AA   | 3          |            |                      | true        |
Dan bevat het resultaat de volgende objecten
| postcode | huisnummer | huisletter | huisnummertoevoeging |
| 1234AA   | 3          |            |                      |

Reply:
Ik vind de tabel inderdaad duidelijker

Je zou ook voor het hele feature de testdata in één keer kunnen definiëren en deze hergebruiken in alle testgevallen. Dan wordt het verschil van resultaten bij verschillende zoekvraag nog duidelijker. Dit kan met Gherkin keyword "Achtergrond".

Nog mooier is het wanneer deze zoekwaarden ook echt in BAG voorkomen, zodat we de feature ook echt kunnen testen en men dit in de praktijk kan zien. Bijvoorbeeld:

```
Achtergrond:
  Gegeven de volgende verzameling van objecten
    | postcode | huisnummer | huisletter | huisnummertoevoeging |
    | 2391PH   | 1          |            |                      |
    | 2391PH   | 1          | A          |  1                   |
    | 2391PH   | 1          | A          |  2                   |
    | 2391PH   | 1          | E          |                      |
    | 2391PH   | 11         |            |                      |
```

Reply:
hier kan je dan meteen alle testgevallen op toepassen (behalve wellicht de situatie van twee adressen met identieke postcode+huisnummer+huisletter+huisnummertoevoeging)

```
| postcode | huisnummer | huisletter | huisnummertoevoeging | exacteMatch | resultaat                                            |
| 2391PH   | 1          |            |                      |             | vindt ook adressen met huisletter of toevoeging      |
| 2391PH   | 1          |            |                      | true        | vindt geen adressen met huisletter of toevoeging     |
| 2391PH   | 1          |            |                      | false       | vindt ook adressen met huisletter of toevoeging      |
| 2391PH   | 1          | A          |                      |             | vindt ook adressen met toevoeging, niet adressen zonder huisletter |
| 2391PH   | 1          | A          |                      | true        | vindt geen adressen zonder huisletter of met toevoeging |
| 2391PH   | 1          | A          | 1                    |             | vindt geen adressen met huisletter of met toevoeging |
| 2391PH   | 1          | A          | 1                    | true        | vindt geen adressen met huisletter of met toevoeging |
| 2391PH   | 2          | A          | 1                    |             | vindt geen adressen                                  |
| 2391PH   | 2          | A          | 1                    | true        | vindt geen adressen                                  |
| 2391PH   | 11         |            |                      |             | vindt alleen het adres met huisnummer 11             |
| 2391PH   | 11         | A          |                      |             | vindt geen adressen                                  |
```